File: flowersec-go/internal/protocolv4/signature_vectors_test.go
package protocolv4

import (
	"bytes"
	"crypto/ed25519"
	"encoding/hex"
	"encoding/json"
	"os"
	"testing"
)

// Primitive interoperability over public fixture secrets only. This does not
// authenticate an issuer, install credentials or qualify a production runtime.
func TestV4SignatureVectors(t *testing.T) {
	raw, err := os.ReadFile("../../../testdata/transport_v4/signatures.json")
	if err != nil {
		t.Fatal(err)
	}
	var corpus struct {
		SchemaSHA string `json:"schema_sha256"`
		Seed      string `json:"signing_seed_hex"`
		Vectors   []struct {
			ID        string `json:"id"`
			Domain    string `json:"domain"`
			Message   string `json:"message_hex"`
			PublicKey string `json:"public_key_hex"`
			Signature string `json:"signature_hex"`
			Accept    bool   `json:"accept"`
		} `json:"vectors"`
	}
	if err := json.Unmarshal(raw, &corpus); err != nil {
		t.Fatal(err)
	}
	if corpus.SchemaSHA != SchemaSHA256 {
		t.Fatal("signature corpus schema drift")
	}
	decode := func(text string) []byte {
		t.Helper()
		result, err := hex.DecodeString(text)
		if err != nil {
			t.Fatal(err)
		}
		return result
	}
	key := ed25519.NewKeyFromSeed(decode(corpus.Seed))
	covered := make(map[string]bool)
	for _, vector := range corpus.Vectors {
		t.Run(vector.ID, func(t *testing.T) {
			message, public, signature := decode(vector.Message), decode(vector.PublicKey), decode(vector.Signature)
			valid := len(public) == ed25519.PublicKeySize && ed25519.Verify(public, message, signature)
			if valid != vector.Accept {
				t.Fatalf("verification=%t, want %t", valid, vector.Accept)
			}
			if vector.Accept {
				if !bytes.Equal(key.Public().(ed25519.PublicKey), public) || !bytes.Equal(ed25519.Sign(key, message), signature) {
					t.Fatal("independent signing differs from the canonical fixture")
				}
				covered[vector.Domain] = true
			}
		})
	}
	var domains []struct {
		Name      string `json:"name"`
		Operation string `json:"operation"`
	}
	if err := json.Unmarshal([]byte(DomainRegistryJSON), &domains); err != nil {
		t.Fatal(err)
	}
	for _, domain := range domains {
		if domain.Operation == "ed25519" && !covered[domain.Name] {
			t.Fatalf("uncovered signature domain %s", domain.Name)
		}
	}
	if !covered[""] {
		t.Fatal("missing external known-answer vector")
	}
}

File: flowersec-ts/src/v4/signatureVectors.test.ts
import { readFileSync } from "node:fs";
import { ed25519 } from "@noble/curves/ed25519.js";
import { describe, expect, it } from "vitest";
import { transportV4Domains, transportV4Registry } from "../generated/transportV4Registry.js";

// Independent primitive consumption of public fixtures; no runtime trust grant.
type Vector = {
  id: string;
  domain: string | null;
  message_hex: string;
  public_key_hex: string;
  signature_hex: string;
  accept: boolean;
};
const corpus = JSON.parse(
  readFileSync(new URL("../../../testdata/transport_v4/signatures.json", import.meta.url), "utf8"),
) as { schema_sha256: string; signing_seed_hex: string; vectors: Vector[] };
const hex = (value: string): Uint8Array => {
  if (!/^(?:[0-9a-f]{2})*$/u.test(value)) throw new Error("invalid fixture hex");
  return Uint8Array.from(Buffer.from(value, "hex"));
};

describe("v4 signature vectors", () => {
  it("binds every Ed25519 domain and an external known answer to the schema", () => {
    expect(corpus.schema_sha256).toBe(transportV4Registry.schemaSHA256);
    const covered = new Set(corpus.vectors.filter((v) => v.accept).map((v) => v.domain));
    expect(covered.has(null)).toBe(true);
    for (const domain of transportV4Domains) {
      if (domain.operation === "ed25519") expect(covered.has(domain.name)).toBe(true);
    }
  });
  for (const vector of corpus.vectors) {
    it(vector.id, () => {
      const message = hex(vector.message_hex), key = hex(vector.public_key_hex), signature = hex(vector.signature_hex);
      const valid = key.length === 32 && signature.length === 64
        && ed25519.verify(signature, message, key, { zip215: false });
      expect(valid).toBe(vector.accept);
      if (vector.accept) {
        const seed = hex(corpus.signing_seed_hex);
        expect(ed25519.getPublicKey(seed)).toEqual(key);
        expect(ed25519.sign(message, seed)).toEqual(signature);
      }
    });
  }
});
